test/test.cpp

```diff
@@ -462,7 +462,114 @@ int do_test_pow(uint64_t aStartPos, int aLabelsCount, unsigned aDiff, unsigned a
+							scryptPositions(cpu_id, s_id, idx_solution, idx_solution, 256, s_salt, SPACEMESH_API_COMPUTE_POW, hash, 512, 1, 1, NULL, NULL, &hashes_computed, &hashes_per_sec);
+							printf("id: ");
+							print_hex32(s_id);
+							printf("\n");
+							printf("salt: ");
+							print_hex32(s_salt);
+							printf("\n");
+							printf("D: ");
+							print_hex32(D);
+							printf("\n");
+							printf("H: ");
+							print_hex32(hash);
+							printf("\n");
+						}
+						break;
+					case SPACEMESH_API_ERROR_NONE:
+						printf("%u hashes, %u h/s, solution not found\n", (uint32_t)hashes_computed, (uint32_t)hashes_per_sec);
+						break;
+					default:
+						printf("error %d, %u hashes, %u h/s\n", status, (uint32_t)hashes_computed, (uint32_t)hashes_per_sec);
+					}
+					if (aSolutionIdx != 0xffffffffffffffffull) {
+						if (idx_solution != aSolutionIdx) {
+							printf("WRONG solution index %ull, expected %ull\n", idx_solution, aSolutionIdx);
+							return 1;
+						}
+					}
+				}
+			}
+
+			return 0;
+		}
+	}
+
+	return 1;
+}
+
+/* test PoW */
+int do_test_leafs_and_pow(uint64_t aStartPos, int labelSize, int aLabelsCount, unsigned aDiff, unsigned aSeed, int aProviderId, uint64_t aSolutionIdx)
+{
+	int providersCount = spacemesh_api_get_providers(NULL, 0);
+
+	if (aSeed) {
+		if (aSeed == -1) {
+			srand(time(nullptr));
+		}
+		else {
+			srand(aSeed);
+		}
+
+		for (int i = 0; i < sizeof(s_id); i++) {
+			s_id[i] = rand();
+		}
+		for (int i = 0; i < sizeof(s_salt); i++) {
+			s_salt[i] = rand();
+		}
+	}
+
+	if (providersCount > 0) {
+		std::unique_ptr<PostComputeProvider> providers_holder((PostComputeProvider *)malloc(providersCount * sizeof(PostComputeProvider)));
+		PostComputeProvider *providers = providers_holder.get();
+
+		if (spacemesh_api_get_providers(providers, providersCount) == providersCount) {
+			uint64_t idx_solution = -1;
+			uint8_t D[32] = { 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff };
+			if (aDiff) {
+				int i = 0;
+				while (aDiff >= 8) {
+					D[i] = 0;
+					i++;
+					aDiff -= 8;
+				}
+				if (aDiff) {
+					D[i] = (1 << (8 - aDiff)) - 1;
+				}
+			}
+			printf("Target D: ");
+			print_hex32(D);
+			printf("\n");
+			uint32_t cpu_id = -1;
+			for (int i = 0; i < providersCount; i++) {
+				if (providers[i].compute_api == COMPUTE_API_CLASS_CPU) {
+					cpu_id = providers[i].id;
+					break;
+				}
+			}
+			for (int i = 0; i < providersCount; i++) {
+				if (providersCount == 1 || providers[i].compute_api != COMPUTE_API_CLASS_CPU)
+				{
+					if (-1 != aProviderId && aProviderId != providers[i].id) {
+						continue;
+					}
+
+					const uint64_t labelsBufferSize = (uint64_t(aLabelsCount) * uint64_t(labelSize) + 7ull) / 8ull;
+					uint8_t *out = (uint8_t *)malloc(labelsBufferSize);
```

Comment:
Given that it's a test that will exit soon, it's not a big deal - I think that `out` is not freed.